### PR TITLE
Add eigenvector centrality function

### PR DESCRIPTION
**Pull request by mtreinish**
This commit adds a new function eigenvector_centrality() to compute the
eigenvector centrality of a graph. It uses the same power function
approach that the NetworkX function eigenvector_centrality() [1]
function uses. This is for two reasons, the first is that a more
traditional eigenvector linear algebra/BLAS function would either
require us to link against BLAS at build time (which is a big change
in the build system and a large requirement) or to call out to numpy
via python both of which seemed less than ideal. The second reason was
to make handling holes in node indices bit easier. Using this approach
also enabled us to put the implementation in retworkx-core so it can be
reused with any petgraph graph.

Part of #441

### CODE REVIEW

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [inline retworkx-core/src/centrality.rs:371]: `n_start` is redundant, we can directly initialize `x`. I removed also the division with `n_start_sum` since my understanding is that NetworkX does it to make sure their initial vector has all entries in [0, 1] but we don't accept a user param for `n_start` as they do, so the all-one vector is fine as a starting choice.

```suggestion
    let mut x: HashMap<G::NodeId, f64> = graph.node_identifiers().map(|n| (n, 1.)).collect();
```

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [inline retworkx-core/src/centrality.rs:370]: I'm wondering if a `Vec` of size `graph.node_bound()` and indexed by [`graph.to_index(n)`](https://docs.rs/petgraph/latest/petgraph/visit/trait.NodeIndexable.html#tymethod.to_index) would be faster for intermediate computations

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [inline retworkx-core/src/centrality.rs:386]: I'm not sure if this is correct in case of parallel edges: `graph.neighbors` will return a neighbor as many times as the number of parallel edges and each time we sum the weights of all parallel edges, so we are "overcounting".

```suggestion
            for edge in graph.edges(*node) {
                let w = weight_fn(edge)?;
                let neighbor = edge.target();
                *x.get_mut(&neighbor).unwrap() += x_last[node] * w;
            }
```

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [inline retworkx-core/src/centrality.rs:391]: If this ever happens it'd mean our current vector is very close to all-zero and a subsequent iteration might not be able to change it much so we'll think that we converged. But convergence to an all-zero solution is not a right solution so I'm thinking we might want to return `None` here.

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [inline retworkx-core/src/centrality.rs:326]:
```suggestion
/// * `weight_fn` - An input callable that will be passed the `EdgeRef` for
```

File: retworkx/__init__.py
Comment (georgios-ts) [inline retworkx/__init__.py:1616]:
```suggestion
    :param graph: Graph to be used. Can either be a
        :class:`~retworkx.PyGraph` or :class:`~retworkx.PyDiGraph`.
```

File: retworkx/__init__.py
Comment (georgios-ts) [inline retworkx/__init__.py:1597]:
```suggestion
    """Compute the eigenvector centrality of a graph.
```

File: retworkx/__init__.py
Comment (georgios-ts) [inline retworkx/__init__.py:1617]:
```suggestion
    :param weight_fn: An optional input callable that will be passed the edge's
```

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:158]:
```suggestion
/// :param PyGraph graph: The graph object to run the algorithm on
```

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:159]:
```suggestion
/// :param weight_fn: An optional input callable that will be passed the edge's
```

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:186]: It might be beneficial to precompute and store all edge weights like we do in `all_pairs_dijkstra_path_lengths` so we avoid calling Python overhead each iteration.

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:220]:
```suggestion
/// :param PyDiGraph graph: The graph object to run the algorithm on
```

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:221]:
```suggestion
/// :param weight_fn: An optional input callable that will be passed the edge's
```

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [inline retworkx-core/src/centrality.rs:363]:
```suggestion
    G: NodeIndexable + IntoNodeIdentifiers + IntoNeighbors + IntoEdges + NodeCount,
    G::NodeId: Eq + std::hash::Hash,
```

File: retworkx-core/src/centrality.rs
Comment (mtreinish) [reply retworkx-core/src/centrality.rs:362]: Oops, yeah this should be 100 not 50:

```suggestion
    let max_iter = max_iter.unwrap_or(100);
```

File: retworkx-core/src/centrality.rs
Comment (mtreinish) [reply retworkx-core/src/centrality.rs:365]: I'll benchmark it's worth testing.

File: retworkx-core/src/centrality.rs
Comment (mtreinish) [reply retworkx-core/src/centrality.rs:381]: Well my thinking was for parallel edges the centrality would be higher so it worked out ok. Similarly if we were going to compute this by generating an adjacency matrix and finding the eigenvector for it we'd be doing a similar thing because the adjacency matrix sums the weight of parallel edges.
'
That being said I don't have a super strong opinion on the correct behavior here, so I'll defer to whatever you think is best

File: retworkx-core/src/centrality.rs
Comment (mtreinish) [reply retworkx-core/src/centrality.rs:386]: That's fine with me, I'll update this: 
```suggestion
        if norm == 0. {
            return Ok(None);
        }
```

File: retworkx-core/src/centrality.rs
Comment (georgios-ts) [reply retworkx-core/src/centrality.rs:378]: Yeah, I agree. It's the right approach to sum the weight of parallel edges but my point was that the total update of the above loop is: `x[neighbor] += x_last[node] * sum(weights of parallel (node, neighbor) edges) * nb. of parallel (node, neighbor) edges` which is incorrect.

File: retworkx-core/src/centrality.rs
Comment (mtreinish) [reply retworkx-core/src/centrality.rs]: I benchmarked it and it did perform better: 

eigv_centrality_times

so I made the change in https://github.com/Qiskit/retworkx/pull/621/commits/09595a14fa9e5ef5526faea5e936fcf976cb6428

File: retworkx-core/src/centrality.rs
Comment (mtreinish) [reply retworkx-core/src/centrality.rs]: Ah, my bad I misread your comment. Yeah that makes sense. I'll correct this

Comment (mtreinish) [reply retworkx-core/src/centrality.rs]: Fixed in: https://github.com/Qiskit/retworkx/pull/621/commits/e03856b1f6ee4430d49f45e915e1683074d54884

File: src/centrality.rs
Comment (mtreinish) [reply src/centrality.rs:186]: Let me benchmark this to see if it makes a difference

Comment (mtreinish) [reply src/centrality.rs:186]: I benchmarked this and it's basically the same but marginally faster in a couple of cases:

eigv_centrality_times_py_callback

not really a noticeable difference. But I think I'll make the change anyway, mostly because it does open up potential parallelization. I think I can make the inner loop execute in parallel and will test that in bit.

Comment (mtreinish) [reply src/centrality.rs]: I pushed this up in: https://github.com/Qiskit/retworkx/pull/621/commits/3204e09eb55509860ead214a116755ec86148637 I'm still waiting on the data collection for the run with a python callback. It's been running the sweep for ~30min and the precomputed weights sweep took ~15min. So there is no doubt this is faster. I'll generate the graph when it eventually finishes just for posterity.

Comment (mtreinish) [reply src/centrality.rs]: Is the graph with a callback set (I just used `float`).

eigv_centrality_times_py_callback

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:205]:
```suggestion
                    if graph.graph.contains_node(NodeIndex::new(k)) {
```

File: src/centrality.rs
Comment (georgios-ts) [inline src/centrality.rs:285]:
```suggestion
                    if graph.graph.contains_node(NodeIndex::new(k)) {
```

**Review by georgios-ts — APPROVED** [review]
Thanks for all the updates. Some minor comments but it's good to go.

**Review by georgios-ts — CHANGES_REQUESTED** [review]
Oh, we'll need to update the docs before we merge.

**Review by georgios-ts — APPROVED** [review]

**Pull request merged**
